connectors/kafka/src/main/java/forklift/controller/KafkaController.java

```diff
@@ -0,0 +1,349 @@
+            }
+        }
+        log.debug("Acknowledge message with topic {} partition {} offset {}", record.topic(), record.partition(), record.offset());
+        return running && this.acknowlegmentHandler.acknowledgeRecord(record);
+    }
+
+    /**
+     * Starts the controller.  Must be called before any other methods.
+     * <p>
+     * <strong>WARNING: </strong>Avoid starting a service which has already run and been stopped.  No attempt is made
+     * to recover a stopped controller.
+     */
+    public void start() {
+        running = true;
+        executor.submit(this::controlLoop);
+    }
+
+    /**
+     * Stops the controller.
+     *
+     * @param timeout  the time to wait for the service to stop
+     * @param timeUnit the TimeUnit of timeout
+     * @throws InterruptedException if the thread is interrupted
+     */
+    public void stop(long timeout, TimeUnit timeUnit) throws InterruptedException {
+        executor.shutdownNow();
+        kafkaConsumer.wakeup();
+        running = false;
+        executor.awaitTermination(timeout, timeUnit);
+    }
+
+    private void controlLoop() {
+        try {
+            while (running) {
+                boolean updatedAssignment = false;
+                if (topics.size() == 0) {
+                    commitPendingAndWaitForTopics();
+                }
+                if (topicsChanged) {
+                    commitPendingOffsetsForRemovedTopics();
+                    kafkaConsumer.subscribe(topics, new RebalanceListener());
+                    updatedAssignment = true;
+                }
+                ConsumerRecords<?, ?> records = flowControlledPoll();
+                //Update the assignment before adding records to stream
+                if (updatedAssignment) {
+                    updateAssignment();
+                }
+                addRecordsToStream(records);
+                Map<TopicPartition, OffsetAndMetadata> offsetData = this.acknowlegmentHandler.getAcknowledged();
+                commitOffsets(offsetData);
+            }
+        } catch (WakeupException e) {
+            log.info("Wakeup, controlLoop exiting");
+        } catch (InterruptedException e) {
+            log.info("Interrupted, controlLoop exiting");
+        } catch (Throwable t) {
+            log.error("controlLoop error, exiting", t);
+            throw t;
+        } finally {
+            running = false;
+            try {
+                Map<TopicPartition, OffsetAndMetadata>
+                                offsetData =
+                                this.acknowlegmentHandler.removePartitions(kafkaConsumer.assignment());
+                try {
+                    if (failedOffset != null) {
+                        log.debug("failedOffset of size: " + failedOffset.size());
+                        for (Map.Entry<TopicPartition, OffsetAndMetadata> entry : failedOffset.entrySet()) {
+                            offsetData.merge(entry.getKey(), entry.getValue(), (oldO, newO) -> {
```

Comment:
Because all of the offsets from the `acknowledgementHandler` are increasing values, can't you just `putAll` the removed offsets (`offsetData`) on top of the `failedOffsets`?